eng/pipelines/build.yml

```
@@ -1,272 +1,299 @@
-      - _InternalInstallArgs:
-          -dotnetruntimeversion '$(DotnetRuntimeVersion)'
-          -dotnetruntimedownloadversion '$(DotnetRuntimeDownloadVersion)'
-          -runtimesourcefeed '$(RuntimeFeedUrl)'
-          -runtimesourcefeedkey '$(RuntimeFeedBase64SasToken)'
-
-    # Only enable publishing in non-public, non PR scenarios.
-    - ${{ if and(ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
-      - _HelixSource: official/dotnet/arcade/$(Build.SourceBranch)
-    - ${{ if or(eq(variables['System.TeamProject'], 'public'), in(variables['Build.Reason'], 'PullRequest')) }}:
-      - _HelixSource: pr/dotnet/arcade/$(Build.SourceBranch)
-
-    # This is only required for cross builds.
-    - ${{ if eq(parameters.crossBuild, true) }}:
-      - _Cross: -cross
-
-    # If there is a native build container, build managed in the host vm/container and native in the nativeBuildContainer
-    - ${{ if ne(parameters.nativeBuildContainer, '') }}:
-      - _ExtraBuildParams: -skipnative
-      - _BuildDisplayName: 'Build Managed'
-
-    # Only add the cross build option if a combined build/test managed/native build (i.e. MacOS arm64)
-    - ${{ if eq(parameters.nativeBuildContainer, '') }}:
-      - _ExtraBuildParams: $(_Cross)
-
-    steps:
-    - ${{ if eq(parameters.osGroup, 'Linux') }}:
+- ${{ each config in parameters.buildConfigs }}:
+  - template: ${{ parameters.jobTemplate }}
+    parameters:
+      name: ${{ coalesce(parameters.name, parameters.osGroup) }}_${{ config.architecture }}_${{ config.configuration }}
+      timeoutInMinutes: ${{ parameters.timeoutInMinutes }}
+      enableMicrobuild: true
+      runAsPublic: ${{ parameters.isCodeQLRun }}
+      templateContext: ${{ parameters.templateContext }}
+
+      pool:
+        # Public Linux Build Pool
+        ${{ if eq(parameters.osGroup, 'Linux') }}:
+          name: $(BuildPool)
+          demands: ImageOverride -equals $(LinuxImage)
+          os: linux
+
+        ${{ elseif eq(parameters.osGroup, 'Windows_NT') }}:
+          name: $(BuildPool)
+          demands: ImageOverride -equals $(WindowsImage)
+          os: windows
+
+        ${{ elseif in(parameters.osGroup, 'MacOS') }}:
+          name: Azure Pipelines
+          vmImage: $(macOSImage)
+          os: macOS
+
+        ${{ elseif and(eq(parameters.osGroup, 'FreeBSD'), ne(variables['System.TeamProject'], 'public')) }}:
+          name: dnceng-freebsd-internal
+          os: freebsd
+
+        ${{ else }}:
+          'Unable to identify pool for required build': error
+
+      ${{ if ne(parameters.container, '') }}:
+        container: ${{ parameters.container }}
+
+      ${{ if ne(parameters.strategy, '') }}:
+        'error, we can no longer support the strategy feature in the new pipeline system. Please remove the strategy from the job template.'
```

should this have an `: error`? Is that how we throw errors in the pipeline jobs?

---

It looks like that is how it is done.  I've never seen it before but Juan added these errors.

---

Can we append `: error` to the end as done on L113 and L143 for consistency?

---

Done

---

Sorry, i meant to make this line 
```suggestion
        'error, we can no longer support the strategy feature in the new pipeline system. Please remove the strategy from the job template.': error
```
not to prefix the other lines with `error, `, as I'm thinking `<string>: error` lets the pipeline know to error out? A guess as I haven't found the docs for that syntax

---

I'm going to leave that for later. I really need to get this in. Can you approve it anyway? Was there anything else?